pkg/link/spoofcheck.go

```diff
@@ -0,0 +1,243 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package link
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/networkplumbing/go-nft/nft"
+	"github.com/networkplumbing/go-nft/nft/schema"
+)
+
+const (
+	natTableName            = "nat"
+	preRoutingBaseChainName = "PREROUTING"
+)
+
+type applyAction func(*nft.Config) error
+type readAction func() (*nft.Config, error)
+
+type SpoofChecker struct {
+	iface      string
+	macAddress string
+	refID      string
+	applyFunc  applyAction
+	readFunc   readAction
+}
+
+func NewSpoofChecker(iface, macAddress, refID string) *SpoofChecker {
+	return &SpoofChecker{iface, macAddress, refID, nft.ApplyConfig, nft.ReadConfig}
+}
+
+func NewSpoofCheckerWithCustomActions(iface, macAddress, refID string, applyF applyAction, readF readAction) *SpoofChecker {
+	sc := NewSpoofChecker(iface, macAddress, refID)
+	sc.applyFunc = applyF
+	sc.readFunc = readF
+	return sc
+}
+
+// Setup applies nftables configuration to restrict traffic
+// from the provided interface. Only traffic with the mentioned mac address
+// is allowed to pass, all others are blocked.
+// The configuration follows the format libvirt and ebtables implemented, allowing
+// extensions to the rules in the future.
+// refID is used to label the rules with a unique comment, identifying the rule-set.
+//
+// In order to take advantage of the nftables configuration change atomicity, the
+// following steps are taken to apply the configuration:
+// - Declare the table and chains (they will be created in case not present).
+// - Apply the rules, while first flushing the iface/mac specific regular chain rules.
+// Note: Although two transactions are taken place, only the 2nd one where the rules
+// are added has a real impact on the system.
+func (sc *SpoofChecker) Setup() error {
+	baseConfig := nft.NewConfig()
+
+	baseConfig.AddTable(&schema.Table{Family: schema.FamilyBridge, Name: natTableName})
+
+	baseConfig.AddChain(sc.baseChain())
+	ifaceChain := sc.ifaceChain()
+	baseConfig.AddChain(ifaceChain)
+	macChain := sc.macChain(ifaceChain.Name)
+	baseConfig.AddChain(macChain)
+
+	if err := sc.applyFunc(baseConfig); err != nil {
+		return fmt.Errorf("failed to setup spoof-check: %v", err)
+	}
+
+	rulesConfig := nft.NewConfig()
+
+	rulesConfig.FlushChain(ifaceChain)
+	rulesConfig.FlushChain(macChain)
+
+	rulesConfig.AddRule(sc.matchIfaceJumpToChainRule(preRoutingBaseChainName, ifaceChain.Name))
+	rulesConfig.AddRule(sc.jumpToChainRule(ifaceChain.Name, macChain.Name))
+	rulesConfig.AddRule(sc.matchMacRule(macChain.Name))
+	rulesConfig.AddRule(sc.dropRule(macChain.Name))
```

Comment:
This is mimicking libvirt chain structure to apply this.
The reason for having another layer in the middle is to allow expanding it to other actions on the same interface. E.g. IP filtering.

Reply:
If we are doing this for future stuff why don't we do it in the future? If will be easier for the upgrade path maybe?

Reply:
Because if one will want to add another action, we do not want to touch this existing logic. Changing the code and keeping the outcome is fine, but if the layers and chain structure change, it presents upgrade and a chain affect on things that are not related (e.g. changing this part logic because someone else adds something new).

I guess with this design, existing experience kicks in and allows the solution to be open for extension and closed to modification.

Reply:
fair enigh

Reply:
I agree with the original question. Why are you jumping to the interface chain just to jump again to the mac chain?
Can you please give an example how it will help to expand it to other actions in the future (for example - how IP filtering will be added?)?

Reply:
The base chain is matching on the interface and redirects it to the interface chain. Currently, there is only one action to filter based on the mac address, but when other actions come into play, they are added to this interface chain.

Here is a simplified example from libvirt with "clean-traffic" rule:
```


        table bridge nat {
                chain PREROUTING {
                        type filter hook prerouting priority dstnat; policy accept;
                        iifname "vnet8" counter packets 0 bytes 0 jump libvirt-I-vnet8
                }
         
                chain libvirt-I-vnet8 {
                        counter packets 0 bytes 0 jump I-vnet8-mac
                        ether type ip counter packets 0 bytes 0 jump I-vnet8-ipv4-ip
                        ether type ip counter packets 0 bytes 0 accept
                        ether type arp counter packets 0 bytes 0 jump I-vnet8-arp-mac
                        ether type arp counter packets 0 bytes 0 jump I-vnet8-arp-ip
                        ether type arp counter packets 0 bytes 0 accept
                        ether type 0x8035 counter packets 0 bytes 0 jump I-vnet8-rarp
                        ether type 0x0835 counter packets 0 bytes 0 accept
                }
         
                chain I-vnet8-mac {
                        ether saddr 52:54:00:86:a7:5e counter packets 0 bytes 0 return
                }
         
                chain I-vnet8-ipv4-ip {
                        ether type ip ip saddr 0.0.0.0 ip protocol udp  counter packets 0 bytes 0 return
                        ether type ip ip saddr 10.0.0.1  counter packets 0 bytes 0 return
                }

                chain I-vnet8-arp-mac {
                        ether type arp #--arp-mac-src 52:54:00:86:a7:5e  counter packets 0 bytes 0 return
                }
         
                chain I-vnet8-arp-ip {
                        ether type arp #--arp-ip-src 10.0.0.1  counter packets 0 bytes 0 return
                }
         
                chain I-vnet8-rarp {
                        ether saddr 52:54:00:86:a7:5e ether daddr ff:ff:ff:ff:ff:ff ether type 0x8035 #--arp-op Request_Reverse --arp-ip-src 0.0.0.0 --arp-ip-dst 0.0.0.0 --arp-mac-src 52:54:00:86:a7:5e --arp-mac-dst 52:54:00:86:a7:5e  counter packets 0 bytes 0 accept
                }
```
Ref for the full dump can be seen [here](https://paste.centos.org/view/30c6a537) (but it is available for the next day only).

Reply:
Thanks for the thorough explanation!